### PR TITLE
Fix of the carrier choice in payment preferences, product editor and cart rules

### LINKED ISSUE #11379: Multiple carriers with same name & delay key will not be display
**Describe the bug**
If two carriers with the same name and transit time (delay) then only one carrier display when we edit product for assign carrier, This is because we are creating the key with name and delay which can be same for multiple carriers. See this code

Code: https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Form/Admin/Product/ProductShipping.php#L71

Some shipping providers will create the same name and delay time carrier like DHL and other. they maintain with the service code of carriers in the different zone.

**To Reproduce**
1. Go to 'Shipping -> Carriers'
2. Click on 'Add Carriers'
3. and add two carriers with the same name & delay (Transit time)
4. Edit any product and see only one carrier display

**Screenshots**
download 2
download 3

**Additionnal information**
PrestaShop version: 1.7.4.4 & 1.7.5.0 beta also
PHP version: 5.6

### PR DESCRIPTION
**Pull request by ghost**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Fix the lack of payment methods when choosing a specific carrier in the FO.
| Type?         | bug fix
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #11379 #16094 
| How to test?  | You only have to create two carriers with the same name, and in the "BO > Payment > Preferences > Carrier restrictions" only ONE will appear, automatically leaving one outside, which will make that when trying to pay in the FO, there will be no payment options available.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/12375)

### CODE REVIEW

**Comment by mickaelandrieu** [comment]
Hi @Rolige,

is the delay always an integer?

I understand why you want to fix it this way but I'm worried about the side effects.

Regards

**Comment by Quetzacoalt91** [comment]
As far as I understand, we would only change the label to make sure all carriers would appear here.

capture d ecran du 2019-01-31 12-04-16

Well, if we have to make sure each carrier is properly differentiated from the others, we could add its identifier.
The result would be:

```php
$choices['#' . $carrier['id_reference'] . ' - ' . $carrier['name']] = $carrier['id_reference'];
```

capture d ecran du 2019-01-31 12-06-04

We need a review from the product team.

**Comment by Rolige** [comment]
Hi,

@mickaelandrieu 
No, it's usually a text.

@Quetzacoalt91 
Include the id_reference, as you comment, would be much more accurate, but it's important include also the "delay" because it's the differentiated with which the administrator user is familiar, as is now done in different sections of PrestaShop, as shown in the following screenshot in the product edition.

image

This is a very critical issue because the selection of payment methods in the FO is completely nulled due to the overwriting generated in the array responsible for generating the list of carriers in the payment preferences.
Continuing with the previous example, I share with you a screenshot of how the transporters are currently shown, as you can see, all those that have the same name now only one appears.

image


Regards!

**Comment by Rolige** [comment]
A video can always show the details better, check here please: https://streamable.com/svqgi

Regards!

**Comment by antho-girard** [comment]
Hello,

Adding the delay is not enough as we may have carriers with the same name **and** the same delay.
We should at least add the ID in the label as suggested by @Quetzacoalt91. And like @Rolige, I think it would be helpful for merchants to have the delay too.

I also notice we have the same behaviour in the Product form if two carriers or more have the same name and the same delay. Adding the carrier ID there too would be great.

**Comment by marionf** [comment]
Hello @Rolige 

It seems to work fine, just if you can add also the ID of the carrier it will be more clear 

Thank you :)

**Comment by Rolige** [comment]
Hi team,

I have updated my patch including the same fix in the payment preferences, product editor and cart rules (Am I forgetting something else?).

Some screenshots of how it would look now...

image
image
image


Regards!

File: classes/CartRule.php [hunk 1]
```diff
@@ -1483,6 +1483,7 @@ public function getAssociatedRestrictions(
 			' . (in_array($type, array('carrier', 'shop')) ? ' AND t.deleted = 0' : '') . '
 			' . ($type == 'cart_rule' ? 'AND t.id_cart_rule != ' . (int) $this->id : '') .
                 $shop_list .
+                ' GROUP BY t.id_' . $type .
```
Comment (Quetzacoalt91) [inline classes/CartRule.php]: Why is this change needed? I wonder if all the carriers will still be displayed.

File: classes/CartRule.php [hunk 2]
```diff
@@ -1497,8 +1498,9 @@ public function getAssociatedRestrictions(
 				' . ($i18n ? 'LEFT JOIN `' . _DB_PREFIX_ . $type . '_lang` tl ON (t.id_' . $type . ' = tl.id_' . $type . ' AND tl.id_lang = ' . (int) Context::getContext()->language->id . ')' : '') . '
 				LEFT JOIN (SELECT id_' . $type . ' FROM `' . _DB_PREFIX_ . 'cart_rule_' . $type . '` WHERE id_cart_rule = ' . (int) $this->id . ') crt ON t.id_' . ($type == 'carrier' ? 'reference' : $type) . ' = crt.id_' . $type . '
 				WHERE 1 ' . ($active_only ? ' AND t.active = 1' : '') .
-                    $shop_list
-                    . (in_array($type, array('carrier', 'shop')) ? ' AND t.deleted = 0' : '') .
+                    $shop_list .
+                    ' GROUP BY t.id_' . $type .
```
Comment (Quetzacoalt91) [inline classes/CartRule.php]: Same comment as the previous GROUP BY

File: classes/CartRule.php (hunk 1 above)
Comment (Rolige) [reply classes/CartRule.php]: Without this, the query return duplicate rows.

File: classes/CartRule.php (hunk 2 above)
Comment (Rolige) [reply classes/CartRule.php]: Same case, without this, the query return duplicate rows.

File: classes/CartRule.php (hunk 1 above)
Comment (Rolige) [reply classes/CartRule.php]: Without this, the query return duplicate rows.

File: classes/CartRule.php (hunk 2 above)
Comment (Rolige) [reply classes/CartRule.php]: Same case, without this, the query return duplicate rows.

File: controllers/admin/AdminCartRulesController.php [hunk 3]
```diff
@@ -640,9 +640,10 @@ public function renderForm()
         $groups = $current_object->getAssociatedRestrictions('group', false, true);
         $shops = $current_object->getAssociatedRestrictions('shop', false, false);
         $cart_rules = $current_object->getAssociatedRestrictions('cart_rule', false, true, 0, $limit);
-        $carriers = $current_object->getAssociatedRestrictions('carrier', true, false);
+        $carriers = $current_object->getAssociatedRestrictions('carrier', true, true);
         foreach ($carriers as &$carriers2) {
             foreach ($carriers2 as &$carrier) {
+                $carrier['name'] = $carrier['id_carrier'].' - '.$carrier['name'].' ('.$carrier['delay'].')';
```
Comment (Quetzacoalt91) [inline controllers/admin/AdminCartRulesController.php]: This line breaks the test made two lines below, it must be moved in the foreach in order to change the value after `PS_SHOP_NAME` is applied

File: classes/CartRule.php (hunk 1 above)
Comment (Quetzacoalt91) [reply classes/CartRule.php]: As I can't reproduce it, I'd like to keep the previous version.

File: classes/CartRule.php (hunk 2 above)
Comment (Quetzacoalt91) [reply classes/CartRule.php]: Same answer, as I can't reproduce it I'd like to keep the previous version.
We can handle this case in another PR if we finally find the issue.

File: controllers/admin/AdminCartRulesController.php (hunk 3 above)
Comment (Rolige) [reply controllers/admin/AdminCartRulesController.php]: Fixed!

File: controllers/admin/AdminCartRulesController.php [hunk 4]
```diff
@@ -640,12 +640,16 @@ public function renderForm()
...
+        $carriers = $current_object->getAssociatedRestrictions('carrier', true, true);
         foreach ($carriers as &$carriers2) {
             foreach ($carriers2 as &$carrier) {
                 foreach ($carrier as $field => &$value) {
-                    if ($field == 'name' && $value == '0') {
-                        $value = Configuration::get('PS_SHOP_NAME');
+                    if ($field == 'name') {
+                        if ($value == '0') {
+                            $value = Configuration::get('PS_SHOP_NAME');
+                        } else {
```
Comment (Quetzacoalt91) [inline controllers/admin/AdminCartRulesController.php:659]:
```suggestion
                        }
```

File: controllers/admin/AdminCartRulesController.php [hunk 5]
```diff
@@ -640,12 +640,16 @@ public function renderForm()
...
+                            $value = $carrier['id_carrier'].' - '.$carrier['name'].' ('.$carrier['delay'].')';
```
Comment (Quetzacoalt91) [inline controllers/admin/AdminCartRulesController.php]:
```suggestion
                        $value = $carrier['id_carrier'].' - '.$carrier['name'].' ('.$carrier['delay'].')';
```

File: controllers/admin/AdminCartRulesController.php [hunk 6]
```diff
@@ -640,12 +640,16 @@ public function renderForm()
...
+                        }
```
Comment (Quetzacoalt91) [inline controllers/admin/AdminCartRulesController.php:664]:
```suggestion
```

**Review by Quetzacoalt91 — DISMISSED** [review]
We still need to apply the the ID and the delay, I remove the surrounding `else`

File: controllers/admin/AdminCartRulesController.php (hunk 4 above)
Comment (Rolige) [reply controllers/admin/AdminCartRulesController.php:659]: You are right, patch applied!

File: classes/CartRule.php (hunk 2 above)
Comment (Rolige) [reply classes/CartRule.php]: The problem with this is that the carriers appears duplicated in the “cart rule > carrier restriction”, this does not occurs in a traditional shop, but in my localhost with multi-lang and multi-store if it happens. Anyway I have changed the fix to the carriers loop taking advantage of the fact that a cycle already exists and this way does not affect the original query, but please check this video about what's happen without this patch in the real life: https://streamable.com/qbdv3

**Comment by marionf** [comment]
@Quetzacoalt91 Could you please have a look as there are new commits ?

File: controllers/admin/AdminCartRulesController.php [hunk 7]
```diff
@@ -640,14 +640,28 @@ public function renderForm()
...
+
                 foreach ($carrier as $field => &$value) {
-                    if ($field == 'name' && $value == '0') {
-                        $value = Configuration::get('PS_SHOP_NAME');
+                    if ($field == 'name') {
+                        if ($value == '0') {
+                            $value = Configuration::get('PS_SHOP_NAME');
+                        }
+
+                        $value = $carrier['id_carrier'].' - '.$carrier['name'].' ('.$carrier['delay'].')';
```
Comment (eternoendless) [inline controllers/admin/AdminCartRulesController.php]: This would replace the previously assigned `$value` on line 657

**Comment by colinegin** [comment]
@Rolige , are you able to make the few changes on your PR that were required ?

File: controllers/admin/AdminCartRulesController.php (hunk 7 above)
Comment (Rolige) [reply controllers/admin/AdminCartRulesController.php]: Fixed!

**Comment by Rolige** [comment]
@colinegin Already fixed!

**Comment by colinegin** [comment]
Amazing ! Thanks a lot ! :)

**Comment by matks** [comment]
Hi @Rolige sorry for the late feedback

I rebased the PR to fix git conflicts and restart CI. I also added https://github.com/PrestaShop/PrestaShop/pull/12379/commits/5c475f21b5a918db1526a05c1cdc3f006dbf9597 to handle usecases where no delay has been provided for carriers.

I'll need one approval from another maintainer to get this merged, but to me it looks good 😄 
@PrestaShop/prestashop-core-developers ?

**Comment by matks** [comment]
I'm still puzzled by this crazy behavior in https://github.com/PrestaShop/PrestaShop/pull/12379/files#diff-bf34b8ed5c52c8da93b3424488bd5047L644 🤔 why is it needed

Also this page is being migrated to Symfony so once it's merged we'll need to re-apply to the Symfony part (I'll do it)

**Comment by Rolige** [comment]
> I'm still puzzled by this crazy behavior in https://github.com/PrestaShop/PrestaShop/pull/12379/files#diff-bf34b8ed5c52c8da93b3424488bd5047L644 🤔 why is it needed
> 
> Also this page is being migrated to Symfony so once it's merged we'll need to re-apply to the Symfony part (I'll do it)

Hi @matks This is needed for the CartRules

**Comment by Rolige** [comment]
@matks In fact, in the recent days I've noticed that a patch is also needed in the carrier preferences, where you can select by best price, best grade or a specific carrier.

**Comment by matks** [comment]
> > I'm still puzzled by this crazy behavior in https://github.com/PrestaShop/PrestaShop/pull/12379/files#diff-bf34b8ed5c52c8da93b3424488bd5047L644 🤔 why is it needed
> > Also this page is being migrated to Symfony so once it's merged we'll need to re-apply to the Symfony part (I'll do it)
> 
> Hi @matks This is needed for the CartRules

What I mean is this strange double foreach, it seems needed but I dont get the full intent 😅 but the code works

I see we got a ✅ from Travis but a 🔴from PrettyCI, I fix that

**Comment by Rolige** [comment]
> > > I'm still puzzled by this crazy behavior in https://github.com/PrestaShop/PrestaShop/pull/12379/files#diff-bf34b8ed5c52c8da93b3424488bd5047L644 🤔 why is it needed
> > > Also this page is being migrated to Symfony so once it's merged we'll need to re-apply to the Symfony part (I'll do it)
> > 
> > 
> > Hi @matks This is needed for the CartRules
> 
> What I mean is this strange double foreach, it seems needed but I dont get the full intent 😅 but the code works
> 
> I see we got a ✅ from Travis but a 🔴from PrettyCI, I fix that

Without it you will have a duplicated results, of course, you need to have more than one language in your installed languages.

Please consider to take a look at my previous comment, this is a related issue also, just I don't have time to review it now.

File: src/PrestaShopBundle/Form/Admin/Product/ProductShipping.php [hunk 8]
```diff
@@ -68,7 +68,12 @@ public function __construct($translator, $legacyContext, $warehouseDataProvider,
         );
         $this->carriersChoices = [];
         foreach ($carriers as $carrier) {
-            $this->carriersChoices[$carrier['name'] . ' (' . $carrier['delay'] . ')'] = $carrier['id_reference'];
+            $choiceId = $carrier['id_carrier'] . ' - ' . $carrier['name'];
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Form/Admin/Product/ProductShipping.php:71]: Duplicate code :thinking: 
Why not using CarrierByReferenceChoiceProvider.php only?

Comment (matks) [reply src/PrestaShopBundle/Form/Admin/Product/ProductShipping.php:71]: It's how the infamous Product page has been built 😭

Comment (PierreRambaud) [reply src/PrestaShopBundle/Form/Admin/Product/ProductShipping.php:71]: :nauseated_face: 
Maybe time to refactor :/

Comment (matks) [reply src/PrestaShopBundle/Form/Admin/Product/ProductShipping.php:71]: Soon 😄

**Review by matks — APPROVED** [review]

**Comment by matks** [comment]
Thank you @Rolige

**Comment by jf-viguier** [comment]
Thanks @Rolige

**Pull request merged**